images/egress/dns-proxy/egress-dns-proxy.sh

```diff
@@ -0,0 +1,183 @@
+#!/bin/bash
+
+# OpenShift egress DNS proxy setup script
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Default DNS nameserver port
+NS_PORT=53
+CONF=/etc/haproxy/haproxy.cfg
+
+BLANK_LINE_OR_COMMENT_REGEX="([[:space:]]*$|#.*)"
+IPADDR_REGEX="[[:digit:]]+\.[[:digit:]]+\.[[:digit:]]+\.[[:digit:]]+"
+PORT_REGEX="[[:digit:]]+"
+DOMAINNAME_REGEX="[[:alnum:]][[:alnum:]-]+?\.[[:alnum:].-]+"
+
+function die() {
+    echo "$*" 1>&2
+    exit 1
+}
+
+function check_prereqs() {
+    if [[ -z "${EGRESS_SOURCE}" ]]; then
+        die "Must specify EGRESS_SOURCE"
+    fi
+    if ! [[ "${EGRESS_SOURCE}" =~ ^${IPADDR_REGEX} ]]; then
+        die "EGRESS_SOURCE must be IPv4 address"
+    fi
+
+    if [[ -z "${EGRESS_DESTINATION}" ]]; then
+        die "Must specify EGRESS_DESTINATION"
+    fi
+}
+
+function validate_port() {
+    local port=$1
+    if [[ "${port}" -lt "1" || "${port}" -gt "65535" ]]; then
+        die "Invalid port: ${port}, must be in the range 1 to 65535"
+    fi
+}
+
+function generate_haproxy_defaults() {
+    echo "
+global
+    log         127.0.0.1 local2
+
+    chroot      /var/lib/haproxy
+    pidfile     /var/run/haproxy.pid
+    maxconn     4000
+    user        haproxy
+    group       haproxy
+
+defaults
+    log                     global
+    mode                    tcp
+    option                  dontlognull
+    option                  tcplog
+    option                  redispatch
+    retries                 3
+    timeout http-request    100s
+    timeout queue           1m
+    timeout connect         10s
+    timeout client          1m
+    timeout server          1m
+    timeout http-keep-alive 100s
+    timeout check           10s
+"
+}
+
+function generate_dns_resolvers() {
+    echo "resolvers dns-resolver"
+    # Fetch nameservers from /etc/resolv.conf
+    declare -a nameservers=$(cat /etc/resolv.conf |grep nameserver|awk -F" " '{print $2}')
+    n=0
+    for ns in ${nameservers[@]}; do
+        n=$(($n + 1))
+        echo "    nameserver ns$n ${ns}:${NS_PORT}"
+    done
+
+    # Add google DNS servers as fallback
```

Comment:
oh no, don't do that. The user may not want to leak their DNS requests to google.